gradle-guide-error-prone/src/main/java/com/palantir/gradle/guide/errorprone/NonAbstractGradleTypeFields.java

```diff
@@ -0,0 +1,160 @@
+import com.google.errorprone.BugPattern.SeverityLevel;
+import com.google.errorprone.VisitorState;
+import com.google.errorprone.bugpatterns.BugChecker;
+import com.google.errorprone.matchers.Description;
+import com.google.errorprone.matchers.Matcher;
+import com.google.errorprone.matchers.Matchers;
+import com.google.errorprone.suppliers.Supplier;
+import com.google.errorprone.suppliers.Suppliers;
+import com.google.errorprone.util.ASTHelpers;
+import com.sun.source.tree.ClassTree;
+import com.sun.source.tree.ExpressionTree;
+import com.sun.source.tree.MethodInvocationTree;
+import com.sun.source.tree.VariableTree;
+import com.sun.tools.javac.code.Symbol.ClassSymbol;
+import com.sun.tools.javac.code.Symbol.VarSymbol;
+import com.sun.tools.javac.code.Type;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import java.util.stream.StreamSupport;
+
+@AutoService(BugChecker.class)
+@BugPattern(severity = SeverityLevel.ERROR, summary = NonAbstractGradleTypeFields.SUMMARY)
+public final class NonAbstractGradleTypeFields extends GradleGuideBugChecker
+        implements BugChecker.VariableTreeMatcher, BugChecker.MethodInvocationTreeMatcher {
+    private static final Matcher<ClassTree> IS_TASK = Matchers.isSubtypeOf("org.gradle.api.Task");
+    private static final Matcher<ExpressionTree> IS_EXTENSION_CREATE = Matchers.instanceMethod()
+            .onDescendantOf("org.gradle.api.plugins.ExtensionContainer")
+            .named("create");
+    private static final Supplier<Type> CLASS_TYPE_SUPPLIER = Suppliers.typeFromString("java.lang.Class");
+    private static final Supplier<Type> PROVIDER_TYPE_SUPPLIER =
+            Suppliers.typeFromString("org.gradle.api.provider.Provider");
+    private static final Supplier<Type> DOMAIN_OBJECT_COLLECTION_TYPE_SUPPLIER =
+            Suppliers.typeFromString("org.gradle.api.DomainObjectCollection");
+    private static final Supplier<Type> FILE_COLLECTION_TYPE_SUPPLIER =
+            Suppliers.typeFromString("org.gradle.api.file.FileCollection");
+
+    public static final String SUMMARY =
+            "Do not declare class fields directly on Tasks or Extensions. Instead, declare an abstract getter "
+                    + "method, e.g., 'public abstract Property<String> getFoo();'. This enables Gradle to inject the "
+                    + "property implementation automatically, removes boilerplate, and supports the Groovy DSL "
+                    + "(e.g. `foo = 3`).";
+
+    @Override
+    public Description matchVariable(VariableTree tree, VisitorState state) {
+        ClassTree enclosingClass = ASTHelpers.findEnclosingNode(state.getPath(), ClassTree.class);
+        if (enclosingClass == null) {
+            return Description.NO_MATCH;
+        }
+
+        if (!IS_TASK.matches(enclosingClass, state)) {
+            return Description.NO_MATCH;
+        }
+
+        if (!enclosingClass.getMembers().contains(tree)) {
+            return Description.NO_MATCH;
+        }
```

Comment:
I think for the enclosingcClass tree/getMembers, instead you can just check if `state.getPath().getParentPath()` is a ClassTree?